File: src/main/scala/imcode/server/document/index/DocumentIndexImpl.scala
package imcode.server.document.index

import java.util.Collections

import _root_.imcode.server.document.DocumentDomainObject
import _root_.imcode.server.document.index.service.{AddDocToIndex, DeleteDocFromIndex, DocumentIndexService}
import _root_.imcode.server.user.UserDomainObject
import com.imcode._
import com.imcode.imcms.ImcmsServicesSupport
import org.apache.solr.client.solrj.SolrQuery

import scala.collection.JavaConverters._


/**
 * {@link DocumentIndex} implementation.
 */
class DocumentIndexImpl(service: DocumentIndexService) extends DocumentIndex with ImcmsServicesSupport with Log4jLogger {

  @deprecated
  override def search(query: DocumentQuery, searchingUser: UserDomainObject): JList[DocumentDomainObject] = {
    val queryString = query.getQuery.toString

    if (logger.isDebugEnabled) {
      logger.debug(s"Searching using *legacy* document query $queryString.")
    }

    val solrQuery = new SolrQuery(queryString)
    for {
      sort <- query.getSort.asOption
      sortField <- sort.getSort
      field <- sortField.getField.asOption
    } {
      solrQuery.addSort(field, if (sortField.getReverse) SolrQuery.ORDER.desc else SolrQuery.ORDER.asc)
      sortField.getReverse
    }

    val documentMapper = imcmsServices.getDocumentMapper

    try {
      val docs = new java.util.LinkedList[DocumentDomainObject]
      for {
        storedDocumentMeta <- search(solrQuery, searchingUser).documentStoredFieldsList().asScala
        doc <- (documentMapper.getDefaultDocument(storedDocumentMeta.id(), storedDocumentMeta.languageCode()): DocumentDomainObject).asOption
      } {
        docs.add(doc)
      }

      docs
    } catch {
      case _: Exception => Collections.emptyList()
    }
  }


  override def search(query: DocumentQuery, searchingUser: UserDomainObject, startPosition: Int, maxResults: Int): com.imcode.imcms.api.SearchResult[DocumentDomainObject] = {
    ???
  }

  @throws(classOf[IndexException])
  override def search(solrQuery: SolrQuery, searchingUser: UserDomainObject): IndexSearchResult = {
    if (solrQuery.get(DocumentIndex.FIELD__LANGUAGE_CODE) == null
      && (solrQuery.getFilterQueries == null ||
      !solrQuery.getFilterQueries.exists(query => query.contains(s"${DocumentIndex.FIELD__LANGUAGE_CODE}:")))) {
      solrQuery.addFilterQuery("%s:%s".format(DocumentIndex.FIELD__LANGUAGE_CODE, imcmsServices.getDocumentLanguages.getDefault.getCode))
    }

    if (solrQuery.get(DocumentIndex.FIELD__META_ID) == null
      && (solrQuery.getFilterQueries == null ||
      !solrQuery.getFilterQueries.exists(query => query.contains(s"${DocumentIndex.FIELD__META_ID}:"))))
      solrQuery.addFilterQuery("%s:%s".format(DocumentIndex.FIELD__META_ID, "[* TO *]"))

    // todo: replace canSearchFor with filter queries
    // UserDomainObject#canSearchFor replacement - not yet complete
    if (!searchingUser.isSuperAdmin) {
      solrQuery.addFilterQuery(s"${DocumentIndex.FIELD__SEARCH_ENABLED}:true")
      solrQuery.addFilterQuery(DocumentIndex.FIELD__ROLE_ID + ":" + searchingUser.getRoleIds.mkString("(", " ", ")"))
    }

    if (solrQuery.getRows == null) solrQuery.setRows(Integer.MAX_VALUE)

    val queryResponse = service.query(solrQuery)
    new IndexSearchResult(solrQuery, queryResponse)
  }


  override def rebuild() {
    service.rebuild()
  }

  override def indexDocument(document: DocumentDomainObject) {
    indexDocument(document.getId)
  }

  override def removeDocument(document: DocumentDomainObject) {
    removeDocument(document.getId)
  }

  override def indexDocument(docId: Int) {
    service.update(AddDocToIndex(docId))
  }

  override def removeDocument(docId: Int) {
    service.update(DeleteDocFromIndex(docId))
  }

  override def getService(): DocumentIndexService = service
}

File: src/main/java/imcode/server/document/index/DocumentIndexImpl.java
package imcode.server.document.index;

import com.imcode.imcms.api.SearchResult;
import com.imcode.imcms.mapping.DocumentMapper;
import imcode.server.Imcms;
import imcode.server.document.DocumentDomainObject;
import imcode.server.document.index.service.AddDocToIndex;
import imcode.server.document.index.service.DeleteDocFromIndex;
import imcode.server.document.index.service.DocumentIndexService;
import imcode.server.document.index.service.UpdateDocumentVersionInIndex;
import imcode.server.user.UserDomainObject;
import lombok.val;
import org.apache.commons.lang3.NotImplementedException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.lucene.search.Sort;
import org.apache.lucene.search.SortField;
import org.apache.solr.client.solrj.SolrQuery;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import static org.springframework.beans.support.PagedListHolder.DEFAULT_PAGE_SIZE;

// translated from scala...
public class DocumentIndexImpl implements DocumentIndex {

    private final static Logger logger = LogManager.getLogger(DocumentIndexImpl.class);
    private final DocumentIndexService service;

    DocumentIndexImpl(DocumentIndexService service) {
        this.service = service;
    }

    /**
     * @deprecated use {@link DocumentIndexImpl#search(org.apache.solr.client.solrj.SolrQuery)}
     */
    @Override
    @Deprecated
    public List<DocumentDomainObject> search(DocumentQuery query, UserDomainObject searchingUser) throws IndexException {
        final String queryString = query.getQuery().toString();

        if (logger.isDebugEnabled()) {
            logger.debug("Searching using *legacy* document query " + queryString);
        }

        final SolrQuery solrQuery = new SolrQuery(queryString);
        solrQuery.setRows(DEFAULT_PAGE_SIZE * 1000); // dummy limit, use SearchDocumentService for better control

        final Sort sort = query.getSort();
        if (sort != null) {
            final SortField[] sortFields = sort.getSort();

            for (SortField sortField : sortFields) {
                final String field = sortField.getField();
                solrQuery.addSort(field, (sortField.getReverse()) ? SolrQuery.ORDER.desc : SolrQuery.ORDER.asc);
            }
        }

        try {
            final DocumentMapper documentMapper = Imcms.getServices().getDocumentMapper();

            return search(solrQuery)
                    .documentStoredFieldsList()
                    .stream()
                    .map(DocumentStoredFields::id)
                    .map(documentMapper::<DocumentDomainObject>getDefaultDocument)
                    .collect(Collectors.toList());

        } catch (Exception e) {
            logger.error(e.getMessage(), e);
            return Collections.emptyList();
        }
    }

    /**
     * @deprecated use {@link DocumentIndexImpl#search(org.apache.solr.client.solrj.SolrQuery)}
     */
    @Override
    @Deprecated
    public SearchResult<DocumentDomainObject> search(DocumentQuery query, UserDomainObject searchingUser, int startPosition, int maxResults) throws IndexException {
        throw new NotImplementedException("Method " + getClass().getName() + "#search SearchResult<DocumentDomainObject> is not implemented");
    }

    @Override
    public void rebuild() throws IndexException {
        service.rebuild();
    }

    @Override
    public void indexDocument(DocumentDomainObject document) throws IndexException {
        indexDocument(document.getId());
    }

    @Override
    public void removeDocument(DocumentDomainObject document) throws IndexException {
        removeDocument(document.getId());
    }

	@Override
	public void updateDocumentVersion(DocumentDomainObject document) throws IndexException {
		updateDocumentVersion(document.getId());
	}

    @Override
    public IndexSearchResult search(SolrQuery solrQuery) throws IndexException {
        val queryResponse = service.query(solrQuery);
        return new IndexSearchResult(solrQuery, queryResponse);
    }

    @Override
    public void indexDocument(int docId) throws IndexException {
        service.update(new AddDocToIndex(docId));
    }

    @Override
    public void removeDocument(int docId) throws IndexException {
        service.update(new DeleteDocFromIndex(docId));
    }

	@Override
	public void updateDocumentVersion(int docId) throws IndexException {
		service.update(new UpdateDocumentVersionInIndex(docId));
	}

    @Override
    public DocumentIndexService getService() {
        return service;
    }

    @Override
    public boolean isUpdateDone() {
        return service.isUpdateDone();
    }
}
